File: 210-course-schedule-ii/course-schedule-ii.ts
function findOrder(numCourses: number, prerequisites: number[][]): number[] {
    if (numCourses === 0) return []
    const indegree: number[] = new Array(numCourses).fill(0)
    const adjList: number[][] = Array.from({ length: numCourses }, () => [])
    const queue: number[] = []
    const result: number[] = []

    for (const [course, prereq] of prerequisites) {
        indegree[course]++
        adjList[prereq].push(course)
    }

    for (let i = 0; i < numCourses; i++) {
        if (indegree[i] === 0) {
            queue.push(i)
            result.push(i)
        }
    }

    while (queue.length > 0) {
        const course: number = queue.shift()
        for (const nextCourse of adjList[course]) {
            indegree[nextCourse]--
            if (indegree[nextCourse] === 0) {
                queue.push(nextCourse)
                result.push(nextCourse)
            }
        }
    }

    return result.length === numCourses ? result : []
}; 

File: 210-course-schedule-ii/course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        indegree = [0] * numCourses
        graph = defaultdict(list)
        order = []

        for course, prereq in prerequisites:
            indegree[course] += 1
            graph[prereq].append(course)

        queue = deque()

        for i in range(numCourses):
            if indegree[i] == 0:
                queue.append(i)

        while queue:
            current = queue.popleft()
            order.append(current)

            for next_course in graph[current]:
                indegree[next_course] -= 1
                if indegree[next_course] == 0:
                    queue.append(next_course)
        
        return order if len(order) == numCourses else []